File: app/src/main/java/org/pdf4664/notepad/activities/TemplateActivity.java
package org.pdf4664.notepad.activities;

import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import org.pdf4664.notepad.utilities.DocUtils;
import org.pdf4664.notepad.dialogs.FileNameDialog;
import org.pdf4664.notepad.utilities.FileUtils;
import org.pdf4664.notepad.R;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.ContextMenu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.Toast;

import java.io.File;

public class TemplateActivity extends AppCompatActivity implements AdapterView.OnItemClickListener {

    private String currentListItem = "";

    final int USE_TEMPLATE = 1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_template);

        Toolbar toolbar = (Toolbar)findViewById(R.id.template_toolbar);
        toolbar.setTitle("Templates");
        setSupportActionBar(toolbar);

        ListView list = (ListView)findViewById(R.id.template_list);
        list.setOnItemClickListener(this);

        loadList();
    }

    private void loadList() {
        String[] contents = DocUtils.templateFileContents(getApplicationContext());

        ListView list = (ListView)findViewById(R.id.template_list);
        ArrayAdapter<String> listContents = new ArrayAdapter<String>(this,
                android.R.layout.simple_expandable_list_item_1,contents);
        list.setAdapter(listContents);
        registerForContextMenu(list);
    }

    private String getListItem(long id) {
        String ret = "";

        ListView fileList = (ListView)findViewById(R.id.template_list);
        for (int pos = 0; pos<fileList.getAdapter().getCount(); pos++) {
            if (fileList.getAdapter().getItemId(pos)==id) {
                ret = fileList.getAdapter().getItem(pos).toString();
            }
        }

        return ret;
    }

    private void copyTemplate(String fileName) {
        File one = new File(DocUtils.templatesPath(getApplicationContext())+currentListItem);
        File two = new File(DocUtils.notesPath(getApplicationContext())+fileName);

        FileUtils.copy(one,two);
    }

    private void useTemplate() {
        Intent intent = new Intent(TemplateActivity.this,FileNameDialog.class);
        startActivityForResult(intent,USE_TEMPLATE);
    }

    private void deleteTemplate() {
        AlertDialog.Builder builder = new AlertDialog.Builder(TemplateActivity.this);
        builder.setTitle("Confirm Delete");
        builder.setMessage("Are you sure you wish to delete this template?");

        builder.setPositiveButton("Delete", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {
                String path = DocUtils.templatesPath(getApplicationContext())+currentListItem;
                File file = new File(path);
                file.delete();

                loadList();
            }
        });

        builder.setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {
                dialogInterface.cancel();
            }
        });

        AlertDialog dialog = builder.create();
        dialog.show();
    }

    @Override
    public void onCreateContextMenu(ContextMenu menu, View v,
                                    ContextMenu.ContextMenuInfo menuInfo) {
        super.onCreateContextMenu(menu,v,menuInfo);
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.template_context_menu,menu);
    }

    @Override
    public boolean onContextItemSelected(MenuItem item) {
        AdapterView.AdapterContextMenuInfo info =
                (AdapterView.AdapterContextMenuInfo)item.getMenuInfo();
        String list_item = getListItem(info.id);
        currentListItem = list_item;

        switch (item.getItemId()) {
            case R.id.delete_template_item: deleteTemplate(); break;
        }
        return true;
    }

    @Override
    public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
        ListView list = (ListView)adapterView;
        currentListItem = list.getItemAtPosition(i).toString();
        useTemplate();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        try {
            String file = data.getExtras().getString("filename");

            switch (requestCode) {
                case USE_TEMPLATE: {
                    copyTemplate(file);
                    finish();
                } break;
            }
        } catch (NullPointerException e) {
        }
    }
}

File: app/src/main/java/org/pdf4664/notepad/activities/TemplateActivity.kt
package org.pdf4664.notepad.activities

import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import org.pdf4664.notepad.utilities.DocUtils
import org.pdf4664.notepad.dialogs.FileNameDialog
import org.pdf4664.notepad.utilities.FileUtils
import org.pdf4664.notepad.R
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.view.ContextMenu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.ListView
import android.widget.Toast

import java.io.File

class TemplateActivity : AppCompatActivity(), AdapterView.OnItemClickListener {

    private var currentListItem = ""

    internal val USE_TEMPLATE = 1

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_template)

        val toolbar = findViewById<View>(R.id.template_toolbar) as Toolbar
        toolbar.title = "Templates"
        setSupportActionBar(toolbar)

        val list = findViewById<View>(R.id.template_list) as ListView
        list.onItemClickListener = this

        loadList()
    }

    private fun loadList() {
        val contents = DocUtils.templateFileContents(applicationContext)

        val list = findViewById<View>(R.id.template_list) as ListView
        val listContents = ArrayAdapter(this,
                android.R.layout.simple_expandable_list_item_1, contents)
        list.adapter = listContents
        registerForContextMenu(list)
    }

    private fun getListItem(id: Long): String {
        var ret = ""

        val fileList = findViewById<View>(R.id.template_list) as ListView
        for (pos in 0 until fileList.adapter.count) {
            if (fileList.adapter.getItemId(pos) == id) {
                ret = fileList.adapter.getItem(pos).toString()
            }
        }

        return ret
    }

    private fun copyTemplate(fileName: String?) {
        val one = File(DocUtils.templatesPath(applicationContext) + currentListItem)
        val two = File(DocUtils.notesPath(applicationContext) + fileName!!)

        FileUtils.copy(one, two)
    }

    private fun useTemplate() {
        val intent = Intent(this@TemplateActivity, FileNameDialog::class.java)
        startActivityForResult(intent, USE_TEMPLATE)
    }

    private fun deleteTemplate() {
        val builder = AlertDialog.Builder(this@TemplateActivity)
        builder.setTitle("Confirm Delete")
        builder.setMessage("Are you sure you wish to delete this template?")

        builder.setPositiveButton("Delete") { dialogInterface, i ->
            val path = DocUtils.templatesPath(applicationContext) + currentListItem
            val file = File(path)
            file.delete()

            loadList()
        }

        builder.setNegativeButton("Cancel") { dialogInterface, i -> dialogInterface.cancel() }

        val dialog = builder.create()
        dialog.show()
    }

    override fun onCreateContextMenu(menu: ContextMenu, v: View,
                                     menuInfo: ContextMenu.ContextMenuInfo) {
        super.onCreateContextMenu(menu, v, menuInfo)
        val inflater = menuInflater
        inflater.inflate(R.menu.template_context_menu, menu)
    }

    override fun onContextItemSelected(item: MenuItem): Boolean {
        val info = item.menuInfo as AdapterView.AdapterContextMenuInfo
        val list_item = getListItem(info.id)
        currentListItem = list_item

        when (item.itemId) {
            R.id.delete_template_item -> deleteTemplate()
        }
        return true
    }

    override fun onItemClick(adapterView: AdapterView<*>, view: View, i: Int, l: Long) {
        val list = adapterView as ListView
        currentListItem = list.getItemAtPosition(i).toString()
        useTemplate()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        try {
            val file = data.extras!!.getString("filename")

            when (requestCode) {
                USE_TEMPLATE -> {
                    copyTemplate(file)
                    finish()
                }
            }
        } catch (e: NullPointerException) {
        }

    }
}
